### ISSUE TITLE
Wrong ESLint plugin loaded (works w/ NPM)

### ISSUE THREAD

**Issue by cmoesel** [yarnpkg/yarn#3332]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
Our project uses _react-scripts_ and _eslint-config-airbnb_ devDependencies.  _react-scripts_ declares _eslint-plugin-import "2.0.1"_ as a dependency.  _eslint-config-airbnb_ declares _eslint-plugin-import "^2.2.0"_ as a peer dependency.  We've added _eslint-plugin-import "^2.2.0"_ as our own devDependency to satisfy the peer dependency.

When we run eslint, we get many of these errors: `error  Definition for rule 'import/no-named-default' was not found`.  Since this rule was added in _eslint-plugin-import v2.2.0_, my best guess is that eslint is looking for the rule in the _eslint-plugin-import v2.0.1_ transitive dependency.

I'm suspecting the issue may be with yarn because if we use `npm install` instead of `yarn install`, then it works as expected and we do not get the previous errors when running eslint.  That said, I'm not confident -- so feel free to disagree.

**If the current behavior is a bug, please provide the steps to reproduce.**
See above.

**What is the expected behavior?**
ESLint should find the rule in _eslint_plugin_import v2.2.0_ rather than _v2.0.1_"

**Please mention your node.js, yarn and operating system version.**
Node 6.6.0, Yarn 0.23.4, Mac OS X 10.12.4

**Comment by cmoesel** [comment]
As noted above, I filed a similar issue against eslint in case the problem is on their end.  Hoping someone from one of the teams might have an idea what is going on.

**Comment by mbifulco** [comment]
Ditto - I'm in the same boat as cmoesel on my project.

**Comment by cmoesel** [comment]
If anyone wants to see this in action, I've created a simple project that reproduces the problem:

https://github.com/cmoesel/eslint-8547

In fact, it's even worse than my real project since there are also some react rules that can't be found (but again, it all works fine if you install with npm instead of yarn).

**Comment by bestander** [comment]
Thanks for a report.
Some more help with investigating what versions are installed in the tree and what is the expected structure will be welcome

**Comment by cmoesel** [comment]
Thanks @bestander.  I've updated my example project with the data for your (and others) convenience.

**NPM**

You will find the results of `npm ls` and the resulting `node_modules` here:
https://github.com/cmoesel/eslint-8547/tree/master/npm_data

In the `npm ls` output, the `eslint-plugin-import` package shows up at the root:
```
├─┬ eslint-plugin-import@2.2.0
│ ├── builtin-modules@1.1.1
│ ├── contains-path@0.1.0
│ ├── doctrine@1.5.0
│ ├─┬ eslint-import-resolver-node@0.2.3
│ │ └─┬ resolve@1.3.3
│ │   └── path-parse@1.0.5
│ ├─┬ eslint-module-utils@2.0.0
│ │ ├─┬ debug@2.2.0
│ │ │ └── ms@0.7.1
│ │ └── pkg-dir@1.0.0
│ ├─┬ has@1.0.1
│ │ └── function-bind@1.1.0
│ ├── lodash.cond@4.5.2
│ ├─┬ minimatch@3.0.4
│ │ └─┬ brace-expansion@1.1.7
│ │   ├── balanced-match@0.4.2
│ │   └── concat-map@0.0.1
│ └─┬ pkg-up@1.0.0
│   └─┬ find-up@1.1.2
│     ├── path-exists@2.1.0
│     └─┬ pinkie-promise@2.0.1
│       └── pinkie@2.0.4
````
and under react-scripts@0.9.5:
```
└─┬ react-scripts@0.9.5
  ├── (bunch of other packages are here...)
  ├─┬ eslint-plugin-import@2.0.1
  │ ├── doctrine@1.3.0
  │ └─┬ eslint-module-utils@1.0.0
  │   └─┬ debug@2.2.0
  │     └── ms@0.7.1
```

**Yarn**

You will find the results of `yarn list` and the resulting `node_modules` here:
https://github.com/cmoesel/eslint-8547/tree/master/yarn_data

In the `yarn list` output, the `eslint-plugin-import` package shows up at the root:
```
├─ eslint-plugin-import@2.2.0
│  ├─ builtin-modules@^1.1.1
│  ├─ contains-path@^0.1.0
│  ├─ debug@^2.2.0
│  ├─ doctrine@1.5.0
│  ├─ eslint-import-resolver-node@^0.2.0
│  ├─ eslint-module-utils@^2.0.0
│  ├─ has@^1.0.1
│  ├─ lodash.cond@^4.3.0
│  ├─ minimatch@^3.0.3
│  └─ pkg-up@^1.0.0
```
and twice (?) under react-scripts@0.9.5:
```
├─ react-scripts@0.9.5
│  ├─ (bunch of other packages are here...)
│  ├─ eslint-plugin-import@2.0.1
│  ├─ eslint-plugin-import@2.0.1
│  │  ├─ builtin-modules@^1.1.1
│  │  ├─ contains-path@^0.1.0
│  │  ├─ debug@^2.2.0
│  │  ├─ doctrine@1.3.x
│  │  ├─ eslint-import-resolver-node@^0.2.0
│  │  ├─ eslint-module-utils@^1.0.0
│  │  ├─ has@^1.0.1
│  │  ├─ lodash.cond@^4.3.0
│  │  ├─ minimatch@^3.0.3
│  │  └─ pkg-up@^1.0.0
```

**Comment by cmoesel** [comment]
Forgot to mention the versions of tools I'm working with:
* Operating System: macOS Sierra 10.12.4
* Node: v6.6.0
* NPM: 3.10.3
* Yarn: 0.23.4

**Comment by bestander** [comment]
So far looks like Yarn is behaving correctly.

The double output of `eslint-plugin-import@2.0.1` is weird but that is ok.

I ran `yarn check --verify-tree` to see if `install` created a correct folder structure from Node.js module resolution point of view.
It passes fine.

And according to your findings both npm and yarn give you correct

```
├─ eslint-plugin-import@2.2.0
├─ react-scripts@0.9.5
│  ├─ eslint-plugin-import@2.0.1
```

I checked the package.json files of all 3 of them and it is all right.

I don't think the issue is here.
Any other ideas?

**Comment by cmoesel** [comment]
Thank you for looking into this, @bestander.  It sounds like both yarn and npm are _technically_ doing the right things from a dependency installation/resolution point of view -- so any difference in eslint behavior is likely due to some variance that is not technically _illegal_ but still somehow affects eslint plugin resolution.

There have been a couple of "me too"s over on the eslint side (see eslint/issues/8547), so I'm hoping that maybe someone over there will be able to take a look and shed some light on this.

Thanks again for taking a look.  I appreciate it.

**Comment by bestander** [comment]
I'll have another look when I have a moment.
Eslint is known for being sensitive to node_modules structure.
Any help to pin point the problem is very much welcome

**Comment by cmoesel** [comment]
See #3535 for what looks like it may be a smoking gun.

**Comment by bestander** [comment]
This is fixed via #3545
